### PR TITLE
[ALLUXIO-2910] WIP: Show Mount Table information if no argument is given to Mount Command

### PR DESCRIPTION
**Pull request by lijiang2**
https://alluxio.atlassian.net/browse/ALLUXIO-2910

### CODE REVIEW

File: conf/alluxio-env.sh.template [hunk 1]
```diff
@@ -1,67 +0,0 @@
-#!/usr/bin/env bash
```
Comment (apc999) [inline conf/alluxio-env.sh.template]: you should not delete this file, independent from your PR

File: conf/alluxio-site.properties.template [hunk 2]
```diff
@@ -1,31 +0,0 @@
-#
```
Comment (apc999) [inline conf/alluxio-site.properties.template]: you should not delete this file, independent from your PR

File: core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java [hunk 3]
```diff
@@ -42,11 +42,13 @@
 import alluxio.exception.status.UnavailableException;
 import alluxio.wire.LoadMetadataType;
 
+import alluxio.wire.MountPointInfo;
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java]: Fix the import order, or the jenkins style checker will be angry
```java
import alluxio.wire.LoadMetadataType;
import alluxio.wire.MountPointInfo;

import org.slf4j.Logger;
```

File: core/client/fs/src/main/java/alluxio/client/file/FileSystem.java [hunk 4]
```diff
@@ -266,6 +268,8 @@ void loadMetadata(AlluxioURI path, LoadMetadataOptions options)
   void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountOptions options)
       throws IOException, AlluxioException;
 
+  Map<String, MountPointInfo> getMountTable() throws IOException, AlluxioException;
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystem.java]: javadoc required

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemMasterClient.java [hunk 5]
```diff
@@ -163,6 +165,8 @@ public static FileSystemMasterClient create(Subject subject, InetSocketAddress m
    */
   void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountOptions options) throws IOException;
 
+  Map<String, MountPointInfo> getMountTable() throws IOException;
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemMasterClient.java]: javadoc required

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java [hunk 6]
```diff
@@ -29,41 +29,7 @@
...
-import alluxio.thrift.MountTResponse;
-import alluxio.thrift.RenameTOptions;
-import alluxio.thrift.RenameTResponse;
-import alluxio.thrift.ScheduleAsyncPersistenceTOptions;
-import alluxio.thrift.ScheduleAsyncPersistenceTResponse;
-import alluxio.thrift.SetAttributeTOptions;
-import alluxio.thrift.SetAttributeTResponse;
-import alluxio.thrift.UnmountTOptions;
-import alluxio.thrift.UnmountTResponse;
+import alluxio.thrift.*;
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java]: let's not use `import alluxio.thrift.*;` but the previous way of importing classes individually to follow Alluxio code convention

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java [hunk 7]
```diff
@@ -300,6 +266,18 @@ public String toString() {
   }
 
   @Override
+  public GetMountTableTResponse getMountTable() throws AlluxioTException {
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java]: 👍 
we need this method to link Thrift service API to the real impl in `DefaultFileSystemMaster`

File: shell/src/main/java/alluxio/shell/command/MountCommand.java [hunk 8]
```diff
@@ -83,6 +83,10 @@ public Options getOptions() {
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
     String[] args = cl.getArgs();
+    if (args.length == 0) {
+      mFileSystem.getMountTable();
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: No print out?

File: core/common/src/thrift/file_system_master.thrift [hunk 9]
```diff
@@ -271,6 +292,9 @@ service FileSystemMasterClientService extends common.AlluxioService {
     )
     throws (1: exception.AlluxioTException e)
 
+  GetMountTableTResponse getMountTable()
```
Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: Put comments here like other APIs. Also can you put this API before `getNewBlockIdForFile` to keep it in order

File: core/common/src/thrift/file_system_master.thrift [hunk 10]
```diff
@@ -123,6 +123,27 @@ struct MountTOptions {
...
 struct MountTResponse {}
 
+struct GetMountTableTResponse {}
+
+/*
+struct GetMountTableTResponse {
+  1: map<string, MountPointInfo> mountTable
+}
+
+struct MountPointInfo {
```
Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: 👍  we need to define `MountPointInfo` in the thrift file. 
the following fields look correct to me

Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: You should change `core/common/src/main/java/alluxio/wire/MountPointInfo.java` to add two new methods `toThrift`. you can check how `FileInfo.java` in the same dir does this.

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java [hunk 11]
```diff
@@ -300,6 +266,18 @@ public String toString() {
...
+    return RpcUtils.callAndLog(LOG, new RpcCallableThrowsIOException<GetMountTableTResponse>() {
+      @Override
+      public GetMountTableTResponse call() throws AlluxioException, IOException {
+          mFileSystemMaster.getMountTable();
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java]: convert the return value from `mFileSystemMaster.getMountTable()` to the thrift format `GetMountTableTResponse`, leveraging `MountPointInfo.toThrift`

**Comment by apc999** [comment]
@lijiang2 I think you build the framework correctly. The missing pieces are

1.  convert the return value 
in `FileSystemMasterClientServiceHandler#getMountTable()` to the thrift data structure `GetMountTableTResponse`, this may require to be able to convert `alluxio.wire.MountPointInfo` to `alluxio.thrift.MountPointInfo`, using a new `toThrift` method following the similar pattern of `alluxio.wire.FileInfo`
2. in `RetryHandlingFileSystemMasterClient#getMountTable()`, `mClient.getMountTable()` will give you the `GetMountTableTResponse`. We should convert this to `Map<String, alluxio.wire.MountPointInfo>`. Note that this may require to implement `fromThrift` method in `alluxio.thrift.MountPointInfo`.
3. in the `MountCommand.java` when `mFileSystem.getMountTable();` returns, you need to print it out based on parsing the returned `Map<String, alluxio.wire.MountPointInfo>`

I also left a few other comments regarding format and javadoc.

File: core/common/src/thrift/file_system_master.thrift (hunk 10 above)
Comment (lijiang2) [inline core/common/src/thrift/file_system_master.thrift]: I wonder if we need to make changes in RetryHandlingFileSystemMasterClient.java.

**Comment by lijiang2** [comment]
I noticed that `alluxio.thrift.MountPointInfo` does not exists now. I wonder if it is generated after we implemented `toThrift` in `alluxio.wire.MountPointInfo` and run `thriftGen`.

**Comment by apc999** [comment]
@lijiang2 in `file_system_master.thrift`, you have definition of `MountPointInfo`.
After you uncomment that part and run `bin/alluxio thriftGen`, it will be created

File: conf/alluxio-env.sh.template (hunk 1 above)
Comment (apc999) [inline conf/alluxio-env.sh.template]: @lijiang2 can you revert the change for this file?
Your PR should not delete this template file.
You can run `git checkout master conf/alluxio-env.sh.template conf/alluxio-site.properties.template` to recover these two files from the master branch

File: core/client/fs/src/main/java/alluxio/client/file/FileSystem.java [hunk 12]
```diff
@@ -267,6 +269,12 @@ void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountOptions options)
       throws IOException, AlluxioException;
 
   /**
+   * Lists all mount points and their corresponding under storage addresses.
+   * @return a table of String and {@link MountPointInfo}
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystem.java]: NIT: a map from String to {@link MountPointInfo}

File: core/client/fs/src/main/java/alluxio/client/file/FileSystemMasterClient.java [hunk 13]
```diff
@@ -164,6 +166,13 @@ public static FileSystemMasterClient create(Subject subject, InetSocketAddress m
   void mount(AlluxioURI alluxioPath, AlluxioURI ufsPath, MountOptions options) throws IOException;
 
   /**
+   * Lists all mount points and their corresponding under storage addresses.
+   *
+   * @return a table of String and {@link MountPointInfo}
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/FileSystemMasterClient.java]: NIT: a map from String to {@link MountPointInfo}

File: core/common/src/thrift/file_system_master.thrift [hunk 14]
```diff
@@ -123,6 +123,27 @@ struct MountTOptions {
 }
 struct MountTResponse {}
 
+//struct GetMountTableTResponse {}
```
Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: remove line 126-128?

File: core/common/src/thrift/file_system_master.thrift [hunk 15]
```diff
@@ -123,6 +123,27 @@ struct MountTOptions {
...
+
+
+struct GetMountTableTResponse {
+  1: map<string, MountPointInfo> mountTable
+}
+
+struct MountPointInfo {
+  // 1: const i64 serialVersionUID = -2912330427506888886L
```
Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: line 134-136, I think you can remove them

File: core/common/src/thrift/file_system_master.thrift (hunk 9 above)
Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: @lijiang2 javadoc?

**Comment by lijiang2** [comment]
Following is the effect of the improvement:

**Comment by apc999** [comment]
@lijiang2 can you revert the change for this file?
Your PR should not delete `conf/alluxio-env.sh.template` and `conf/alluxio-site.properties.template`.
You can run 
```bash
$ git checkout master conf/alluxio-env.sh.template conf/alluxio-site.properties.template to recover 
```
these two files from the master branch

File: core/common/src/thrift/file_system_master.thrift [hunk 16]
```diff
@@ -123,6 +123,21 @@ struct MountTOptions {
...
+  1: string ufsUri
+  2: string ufsType
+  3: i64 ufsCapacityBytes = -1
+  4: i64 ufsUsedBytes = -1
+  5: bool readOnly
+  6: map<string, string> properties
+  7: bool shared
+}
+
+
```
Comment (apc999) [inline core/common/src/thrift/file_system_master.thrift]: NIT: one extra empty line

File: core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java [hunk 17]
```diff
@@ -66,13 +67,18 @@
 import alluxio.thrift.UnmountTResponse;
 import alluxio.wire.ThriftUtils;
 
+import alluxio.wire.MountPointInfo;
+
 import com.google.common.base.Preconditions;
+
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientServiceHandler.java]: NIT: remove empty line here

File: shell/src/main/java/alluxio/shell/command/MountCommand.java [hunk 18]
```diff
@@ -16,12 +16,15 @@
 import alluxio.client.file.options.MountOptions;
 import alluxio.exception.AlluxioException;
 
+import alluxio.wire.MountPointInfo;
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: NIT: remove line 18

File: shell/src/main/java/alluxio/shell/command/MountCommand.java [hunk 19]
```diff
@@ -83,6 +86,21 @@ public Options getOptions() {
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
     String[] args = cl.getArgs();
+    if (args.length == 0) {
+      Map<String, MountPointInfo> mountTable = mFileSystem.getMountTable();
+      String leftAlignFormat = " %-20s  %-3s %-60s %-8s %-15d %-15d %-5b %-5b   ";
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: NIT: define this as a static constant in class `MountCommand`

Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: why we need trailing white space after `%-5b`?

File: shell/src/main/java/alluxio/shell/command/MountCommand.java [hunk 20]
```diff
@@ -83,6 +86,21 @@ public Options getOptions() {
   @Override
   public int run(CommandLine cl) throws AlluxioException, IOException {
     String[] args = cl.getArgs();
+    if (args.length == 0) {
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: can you also update the `public String getUsage()` method, to have a second line to show the usage of no arg in addition to the one with two args

Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: can you also update the `public String getUsage()` method, to have a second line to show the usage of no arg in addition to the one with two args

Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: can you also update the `public String getUsage()` method, to have a second line to show the usage of no arg in addition to the one with two args

File: shell/src/main/java/alluxio/shell/command/MountCommand.java [hunk 21]
```diff
@@ -83,6 +86,21 @@ public Options getOptions() {
...
+      Map<String, MountPointInfo> mountTable = mFileSystem.getMountTable();
+      String leftAlignFormat = " %-20s  %-3s %-60s %-8s %-15d %-15d %-5b %-5b   ";
+      for (Map.Entry<String, MountPointInfo> entry :
+              mountTable.entrySet()) {
+        String mMountPoint = entry.getKey();
+        MountPointInfo mountPointInfo = entry.getValue();
+        System.out.format(leftAlignFormat, mMountPoint, "on", mountPointInfo.getUfsUri(),
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: I think it should be `mountPointInfo.getUfsUri(), "on", mMountPoint`,
for the following attributes, can you make the output look like:
"(local, capacity=xxx bytes, used bytes=xxxx bytes, not read-only, shared)"

**Comment by apc999** [comment]
@lijiang2  thanks for the big improvement. I left a bunch comments inline (please see https://github.com/Alluxio/alluxio/pull/5794/files)

- can you re-run 
```bash
$ bin/alluxio thriftGen
```
as I am not sure why `GetPinnedFileIdsTResponse.java`, `PersistCommandOptions.java` and a few others are updated, we didn't update their thrift definition in this PR.

- also in `core/common/src/thrift/file_system_master.thrift`, can you put a paragraph to comment the API `GetMountTableTResponse getMountTable()`?

File: shell/src/main/java/alluxio/shell/command/MountCommand.java (hunk 19 above)
Comment (lijiang2) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: Because I designed the last `%-5b` corresponds to `shared` and following is the `properties` map printed by `println`. I was not sure how to print map using formatter, so I decided to use `println` in the end.

File: shell/src/main/java/alluxio/shell/command/MountCommand.java (hunk 21 above)
Comment (lijiang2) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: I was wondering where to print the `properties` map.

**Comment by lijiang2** [comment]
Just re-ran the thriftGen and no files are changed after the re-run.

File: shell/src/main/java/alluxio/shell/command/MountCommand.java [hunk 22]
```diff
@@ -97,15 +115,15 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
...
     mFileSystem.mount(alluxioPath, ufsPath, options);
     System.out.println("Mounted " + ufsPath + " at " + alluxioPath);
     return 0;
   }
 
   @Override
   public String getUsage() {
-    return "mount [--readonly] [--shared] [--option <key=val>] <alluxioPath> <ufsURI>";
+    return "mount [--readonly] [--shared] [--option <key=val>] <alluxioPath> <ufsURI>\n"
+            + "mount [--readonly] [--shared] [--option <key=val>]";
```
Comment (apc999) [inline shell/src/main/java/alluxio/shell/command/MountCommand.java]: just "mount". the part of "[--readonly] [--shared] [--option <key=val>]" does not make sense when mounting without args
